### DISCUSSION TITLE
Ability to run script a script and save result into JSON at build step

### DISCUSSION THREAD

**Discussion by Stranget0** [withastro/roadmap#562] [Proposal]
### Body

# Summary

Say you are fetching some data and you're using it in client side script somewhere. Similar to fetching and building HTML from it, but this could be used by JS.

# Background & Motivation

I've been building a script heavy application. I could make it fetch the data and save it in `data-` attributes on some element, only to get it from element later, but why not have it in JSON instead if it is complex? This RFC feels like a nice bridge.

# Goals

- A way of defining build-step script that saves its result into a specified file
- An easier middleground between template generation and client script

# Example

`somefile.js`, resolves into `somefile.json`
```js
// Async or sync for complex calculations
export default async function buildStepScript() {
  const data = await fetchSomeData();
  return data;
}
```
You are then able to use it in your script tag:

`index.astro`
```html
<script>
  import data from "@/preloaded/somefile.json";
  data.forEach(console.log);
</script>
```

Also we could make this typesafe by using zod and declaring a module. What are your thoughts?

**Comment by delucis** [comment]
Creating a JSON file is possible today using an endpoint: https://docs.astro.build/en/core-concepts/endpoints/#static-file-endpoints

But looking at the `import`, it sounds like you also want to _bundle_ the JSON with your script tag? In that case you may be able to use a Vite virtual module plug-in. The plug-in could create a virtual module containing the JSON you fetched and then script tags could import and bundle it.

**Reply by Stranget0** [reply]
Thank you!
